### ISSUE TITLE
[loader] Switch dlopen's default scope from RTLD_GLOBAL to RTLD_LOCAL

### ISSUE THREAD

**Issue by CoffeeFlux** [mono/mono#17606]
Currently, when calling dlopen we default to RTLD_GLOBAL, and the embedding API is designed with that in mind.

See: https://github.com/mono/mono/blob/e74736a471f50781db06914f1c048aadabfb52bc/mono/utils/mono-dl-posix.c#L135

This appears to have been the case since the initial implementation of mono_dl, and the reason for the choice is probably lost to time. Unfortunately, this creates an inconsistency with Windows since it has no such concept, and can also break some libraries. Conversely, using RTLD_LOCAL can also break libraries that rely on global.

As a result of all this, CoreCLR is explicit about defaulting to RTLD_LOCAL, and we should really match their behavior at least on netcore. We probably want to preserve the embedding API behavior and just change the runtime so that calls from internal or managed APIs will default to local. Changing the embedding API would not only require an API change, but there is a significant chance of it breaking customer applications.

I'm still not sure whether I can safely change the behavior for internal/managed on legacy Mono. My inclination is no, just do it on netcore, but I would welcome other opinions on the matter. Starting with both profiles and scaling it back if integration tests fail is also an option. Changing the embedding API behavior for netcore only is also something to consider, but I'm not particularly fond of the idea.

tl;dr Internal dlopen use on netcore needs to be RTLD_LOCAL, embedding API use probably needs to be RTLD_GLOBAL, internal use on legacy is up in the air.

**Comment by CoffeeFlux** [comment]
Maybe relevant: https://github.com/xamarin/xamarin-android/commit/d9136d5ccea547773254d2aa21c75ad88f42445d

**Comment by CoffeeFlux** [comment]
@grendello indicated it's probably fine to change internally on legacy from Android's perspective. Will ask the Unity folks, and if they aren't concerned about it I'll try changing it for everything and see what happens with integration.

**Comment by jaykrell** [comment]
The connection to Windows is tenuous.
Windows does not offer either of the Unix semantics. Symbols are qualified by dll, and dlls are in a global space once loaded.

**Comment by CoffeeFlux** [comment]
Gotcha. Still, the symbols aren't going to all just be put directly into a global namespace, which is what I assume Jan meant by [this](https://github.com/dotnet/coreclr/pull/18628#issuecomment-399922248), so I still agree with CoreCLR that local is the best default here.
